### ISSUE TITLE
about  train

### ISSUE THREAD

**Issue by LawrenceWangzi** [wvangansbeke/LaneDetection_End2End#17]
about train
could i ask some question?
tusimple dataset only include some images  and *.json files .what is --image_dir and --gt_dir?(run code python main.py --image_dir /path/to/image/folder --gt_dir /path/to/ground_truth/folder --end_to_end True)

**Comment by wvangansbeke** [comment]
Hi @LawrenceWangzi,

You're correct. Tusimple only includes json files and rgb images. However, if you want to do supervised segmentation, you'll need pixelwise ground truth maps. You can download these via the download link provided in Readme. I made these myself based on the json files. You need to point to this directory using the --gt_dir flag. However, the ground truth is actually only required when you perform vanilla segmentation which is not end-to-end. To perform end-to-end lane detection you can use the --end-to-end flag.

Hope this helps. 
Wouter

**Comment by LawrenceWangzi** [comment]
@wvangansbeke thank very much！I could not understand  you words.Could you send the gt_dir files to me（2274208622@qq.com）.I just want to understand the format of the gt_dir files .
yours
Lawrence

**Comment by wvangansbeke** [comment]
Hi Lawrence,

You can find de download link for the dataset in the README file. It already contains the ground truth files. You need to point to this locations with the --gt_dir argument. 

Kind regards, 
Wouter

**Comment by ryan1051** [comment]
> Hi @LawrenceWangzi,
> 
> You're correct. Tusimple only includes json files and rgb images. However, if you want to do supervised segmentation, you'll need pixelwise ground truth maps. You can download these via the download link provided in Readme. I made these myself based on the json files. You need to point to this directory using the --gt_dir flag. However, the ground truth is actually only required when you perform vanilla segmentation which is not end-to-end. To perform end-to-end lane detection you can use the --end-to-end flag.
> 
> Hope this helps.
> Wouter

Hi, do you mean that, if I set the --end-to-end flag True, I don't need to set the --gt_dir argument?
Just python main.py --image_dir /path/to/image/folder  --end_to_end True ?
Thank you.

**Comment by wvangansbeke** [comment]
You're correct there is no need for the ground truth segmentation when `end_to_end` is true. However, in my dataloader I always include the ground truth segmentation maps. You can easily comment out that part if you don't need it.

Kind regards,
Wouter

**Comment by ryan1051** [comment]
@wvangansbeke 
thank you. And how do you get the ground truth image(.png), using script transform .json to .png?
Can you show me the script?

**Comment by wvangansbeke** [comment]
Hi @LawrenceWangzi,

That's actually quite simple. I just drew a line through the points provided in the json files. Opencv has a lot of functions to draw on images. I think I used `cv2.polylines()` to draw the lines. You just have to specify the thickness of the line.

**Comment by ryan1051** [comment]
Thank you, I will try it.
Another question, if I set --end-to-end flag True, how dose it complete the train task only using the --image_dir /path/to/image/folder, without any labels?

**Comment by wvangansbeke** [comment]
You'll have to comment out some lines in the dataloader where I load the ground truth segmentation maps. (= the variable `gt` in `Dataloader/Load_Data_new.py`).

**Comment by ryan1051** [comment]
@wvangansbeke  I find the subset of TuSimple you used has 2535 items, and the label_data_all.json and Curve_parameters.json both have 3626 items. Can I train directly using the images and jsons? And why the number different?

**Comment by ryan1051** [comment]
and what the label_new.json represents?

**Comment by wvangansbeke** [comment]
Hi @ryan1051,

You are correct. I used a subset of 2535 images but I computed the parameters for the whole dataset. I left some images out for training were the car is switching lanes, since you'll need an instance segmentation to handle those cases. Furthermore, some ground truth coordinates for the lane lines are ordered differently in the json files. So you will have to check the order manually. That's why some images were left out, since the order might be incorrect.

`Label_new.json` contains the position of the car on the road. How many lanes are left and right of the car? It is composed relative to the car. Imagine that the car is situated in the middle of the array. A `1` on the left from the middle, means that a lane is present on the left. A `1` on the right means the a lane is present on the right and so on. I did this for every image. 

Hope it is clear, 
Wouter

**Comment by wvangansbeke** [comment]
I wanted to add something. The `Label_new.json` file also contains the line types relative to the car. A `1` means the line is a continuous line, a `0` means it is a dashed line and a `-1` means the line is not present. The latter is really important, because the network will always output 2 or 4 lane lines. Sometimes there are less lanes on the road after all. Furthermore, in this way I was able to colorize the lines in the visualization. 

Kind regards,
Wouter

**Comment by whlsper** [comment]
hi,
i want to ask,can you provide the handled gt_dir?
i mean the gt_dir(.png) with haved drawn done。

**Comment by wvangansbeke** [comment]
Hi @whlsper,

I see the misunderstanding. The ground truth should be fine. Since the maximum value of the png is 255 and the values of the lane line range between 0 and 4, they are just hard to see. I could have used a colormap in order to prevent this.

Best,
Wouter

**Comment by whlsper** [comment]
> Hi @whlsper,
> 
> I see the misunderstanding. The ground truth should be fine. Since the maximum value of the png is 255 and the values of the lane line range between 0 and 4, they are just hard to see. I could have used a colormap in order to prevent this.
> 
> Best,
> Wouter

Hi @wvangansbeke 

I want to ask you another question,when i use the dataset you provide to train,there has a problem like 

"1it [00:07,  7.19s/it]Batch with idx [  90 1861  619 2277 1481  682 2498  671  856  182 1472 1288  644 2089
 1379 2642 1724  747 2430 1153  900 1436  348 1041 1609  284 1156  405
 1764 2417   68 2000 2014  370  131 2519  302 1811 1682  502 2108  285
 2429 2017 2552 2076 1221 1691  709   64  336 1376 2212 2051  809 2133
  978  724   44 2581  653 1012  720  869] skipped due to singular matrix"

Can you help me?

**Comment by wvangansbeke** [comment]
Hi @whlsper,

Which ground truth are you using: segmentation ground truth or the line coefficients (look at the `end_to_end` variable)? Maybe remove the try - catch statement to see which error you get exactly.

Kind regards,
Wouter
